pyquil/noise.py

```diff
@@ -358,467 +360,989 @@ def damping_after_dephasing(T1: float, T2: float, gate_time: float) -> List[np.n
+	"""
+	from pyquil.quil import Program
+
+	if p0 is None:  # pragma no coverage
+		p0 = Program()
+
+	p_i = (
+			p0
+			+ Program(
+		Declare("ro", "BIT", 1),
+		I(q),
+		MEASURE(q, MemoryReference("ro", 0)),
+	)
+	).wrap_in_numshots_loop(trials)
+	results_i = np.sum(_run(qc, p_i))
+
+	p_x = (
+			p0
+			+ Program(
+		Declare("ro", "BIT", 1),
+		RX(np.pi, q),
+		MEASURE(q, MemoryReference("ro", 0)),
+	)
+	).wrap_in_numshots_loop(trials)
+	results_x = np.sum(_run(qc, p_x))
+
+	p00 = 1.0 - results_i / float(trials)
+	p11 = results_x / float(trials)
+	return np.array([[p00, 1 - p11], [1 - p00, p11]])
+
+
+def _run(qc: "PyquilApiQuantumComputer", program: "Program") -> List[List[int]]:
+	result = qc.run(qc.compiler.native_quil_to_executable(program))
+	bitstrings = result.readout_data.get("ro")
+	assert bitstrings is not None
+	return cast(List[List[int]], bitstrings.tolist())
+
+
+single_qubit_noise_1Q_gate_name = "Damping_after_dephasing_for_1Q_gate"
+single_qubit_noise_2Q_gate_name = "Damping_after_dephasing_for_2Q_gate"
+Depolarizing_1Q_gate = "1Q_gate"
+Depolarizing_pre_name = "Depolarizing_"
+No_Depolarizing = "No_Depolarizing"
+
+
+def get_qc_name(qc: QuantumComputer):
+	"""
+	returns the name of the quantum computer `qc`,
+	without the ending 'qvm' if it exists.
+	"""
+	name = qc.name
+	if name[-4:] == "-qvm":
+		name = name[0:-4]
+		return name
+	else:
+		return name
+
+
+def change_times_by_ratio(times: Dict[Any, float], ratio: float):
+	"""
+	change the times in the dict `times` by a given `ratio`.
+	larger ratio will make the times shorter (i.e. more noise)
+	"""
+	for key in times.keys():
+		times[key] = times[key] / (ratio + 1e-10)  # no zerp devision
+	return times
+
+
+class Calibrations:
+	"""
+    encapsulate the calibration data for Aspen-M-2 or Aspen-M-3 machine.
+
+	contains:
+		T1
+		T2
+		fidelities: 1Q_gate, and a dict for each 2Q gate
+		readout
+
+    args: qc (QuantumComputer, optional): a Quantum Computer (Aspen-M-2 or Aspen-M-3).
+    Defaults to None, where the user can define his own calibration data.
+
+	Notice: this class heavily relies on the specific way on which the lattices are written.
+	this may change in time, and require changes in the class.
+    """
+
+	def __init__(self, qc: Optional[QuantumComputer] = None, cal=None):
+		self.fidelities = {}
+		self.readout_fidelity = {}
+		self.T2 = {}
+		self.T1 = {}
+		self.two_q_gates = set()
+
+		if cal is not None:
+			self.T1 = cal.T1.copy()
+			self.T2 = cal.T2.copy()
+			self.readout_fidelity = cal.readout_fidelity.copy()
+			self.fidelities = {}
+			for key, val in cal.fidelities.items():
+				self.fidelities[key] = val.copy()
+			self.two_q_gates = cal.two_q_gates
+			return
+
+		if qc is None:
+			return  # user can set his own values
+
+		else:
+			qc_name = get_qc_name(qc)
+		if qc_name not in ["Aspen-M-2", "Aspen-M-3"]:
+			raise ValueError("qc must be Aspen-M-2 or Aspen-M-3")
+		else:
+			url = "https://forest-server.qcs.rigetti.com/lattices/"
```

Comment:
The ISA is actually already accessible on a `QuantumProcessor` (which is a property of a `QuantumComputer`).

For example, you can easily get 1Q and 2Q fidelities from the compiler ISA:

```py
from pyquil.quantum_processor.qcs import get_qcs_quantum_processor

qpu = get_qcs_quantum_processor("Aspen-M-3")

isa = qpu.to_compiler_isa()

isa.qubits

# Qubit(
#     id=0,
#     dead=False,
#     gates=[
#         GateInfo(operator="RX", duration=50.0, fidelity=1.0, parameters=[0.0], arguments=[0], operator_type="gate"),
#         GateInfo(
#             operator="RX",
#             duration=50.0,
#             fidelity=0.9953775310422732,
#             parameters=[3.141592653589793],
#             arguments=[0],
#             operator_type="gate",
#         ),
#         GateInfo(
#             operator="RX",
#             duration=50.0,
#             fidelity=0.9953775310422732,
#             parameters=[-3.141592653589793],
#             arguments=[0],
#             operator_type="gate",
#         ),
#         GateInfo(
#             operator="RX",
#             duration=50.0,
#             fidelity=0.9953775310422732,
#             parameters=[1.5707963267948966],
#             arguments=[0],
#             operator_type="gate",
#         ),
#         GateInfo(
#             operator="RX",
#             duration=50.0,
#             fidelity=0.9953775310422732,
#             parameters=[-1.5707963267948966],
#             arguments=[0],
#             operator_type="gate",
#         ),
#         GateInfo(operator="RZ", duration=0.01, fidelity=1.0, parameters=["_"], arguments=[0], operator_type="gate"),
#         MeasureInfo(operator="MEASURE", duration=2000.0, fidelity=0.952, qubit=0, target="_", operator_type="measure"),
#         MeasureInfo(operator="MEASURE", duration=2000.0, fidelity=0.952, qubit=0, target=None, operator_type="measure"),
#     ],
# )

isa.edges.get("20-123")

# Edge(
#     ids=[20, 123],
#     dead=False,
#     gates=[
#         GateInfo(
#             operator="XY",
#             duration=200.0,
#             fidelity=0.9637310922221702,
#             parameters=["theta"],
#             arguments=["_", "_"],
#             operator_type="gate",
#         )
#     ],
# )
```

The raw data for the ISA, which you'll need to access T1 and T2 data, has a less obvious structure, but if you know the benchmarks you are looking for, you can get what you need. For example:

```python
def get_t1s(qpu) -> Dict[int, float]:
    benchmarks = qpu._isa.benchmarks
    operations = next(operation for operation in benchmarks if operation.name == "FreeInversionRecovery")
    t1s = {site.node_ids[0]: site.characteristics[0].value for site in operations.sites}
    return t1s


def get_t2s(qpu) -> Dict[int, float]:
    benchmarks = qpu._isa.benchmarks
    operations = next(operation for operation in benchmarks if operation.name == "FreeInductionDecay")
    t1s = {site.node_ids[0]: site.characteristics[0].value for site in operations.sites}
    return t1s

qpu = get_qcs_quantum_processor("Aspen-M-3")
t1s = get_t1s(qpu)
t2s = get_t2s(qpu)
```

Reply:
Great! I'll use it.